pkg/northbound/admin/admin_test.go

```diff
@@ -0,0 +1,182 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package admin
+
+import (
+	"context"
+	"fmt"
+	"github.com/atomix/atomix-go-client/pkg/atomix/test"
+	"github.com/atomix/atomix-go-client/pkg/atomix/test/rsm"
+	"github.com/onosproject/onos-api/go/onos/config/admin"
+	devicechange "github.com/onosproject/onos-api/go/onos/config/change/device"
+	"github.com/onosproject/onos-api/go/onos/config/device"
+	devicesnapshot "github.com/onosproject/onos-api/go/onos/config/snapshot/device"
+	networkchanges "github.com/onosproject/onos-config/pkg/store/change/network"
+	devicesnapshotstore "github.com/onosproject/onos-config/pkg/store/snapshot/device"
+	networksnapshotstore "github.com/onosproject/onos-config/pkg/store/snapshot/network"
+	"github.com/stretchr/testify/assert"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/test/bufconn"
+	"io"
+	"net"
+	"testing"
+)
+
+func setUpServer(ctx context.Context, t *testing.T) (*grpc.ClientConn, admin.ConfigAdminServiceClient, *grpc.Server, devicesnapshotstore.Store) {
+	lis := bufconn.Listen(1024 * 1024)
+	s := grpc.NewServer()
+
+	atomixTest := test.NewTest(
+		rsm.NewProtocol(),
+		test.WithReplicas(1),
+		test.WithPartitions(1))
+	assert.NoError(t, atomixTest.Start())
+
+	atomixClient, err := atomixTest.NewClient("node")
+	assert.NoError(t, err)
+
+	networkChangesStore, err := networkchanges.NewAtomixStore(atomixClient)
+	assert.NoError(t, err)
+
+	networkSnapshotStore, err := networksnapshotstore.NewAtomixStore(atomixClient)
+	assert.NoError(t, err)
+
+	deviceSnapshotStore, err := devicesnapshotstore.NewAtomixStore(atomixClient)
+	assert.NoError(t, err)
+
+	admin.RegisterConfigAdminServiceServer(s, &Server{
+		networkChangesStore:  networkChangesStore,
+		networkSnapshotStore: networkSnapshotStore,
+		deviceSnapshotStore:  deviceSnapshotStore,
+	})
+
+	go func() {
+		if err := s.Serve(lis); err != nil {
+			t.Error("Server exited with error")
+		}
+	}()
+
+	dialer := func(ctx context.Context, address string) (net.Conn, error) {
+		return lis.Dial()
+	}
+
+	conn, err := grpc.DialContext(context.Background(), "bufnet", grpc.WithContextDialer(dialer), grpc.WithInsecure())
+	if err != nil {
+		t.Error("Failed to dial bufnet")
+	}
+
+	client := admin.CreateConfigAdminServiceClient(conn)
+	assert.NoError(t, err)
+
+	go func() {
+		<-ctx.Done()
+		_ = atomixTest.Stop()
+		s.Stop()
+		conn.Close()
+	}()
+
+	return conn, client, s, deviceSnapshotStore
+}
+
+func Test_RollbackNetworkChange_BadName(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	conn, client, server, _ := setUpServer(ctx, t)
+	defer server.Stop()
+	defer conn.Close()
```

Comment:
These deferrals are unnecessary since the setup function is already calling them when the context is closed. We should only have to cancel the context, so we can actually remove these lines from the test functions. Then, we may not need to return the client and server from the setup function either. Deleting code is always good, and I guess this is one of the benefits of the context pattern…